Gifski/GifskiWrapper.swift

```diff
@@ -68,18 +68,45 @@ final class GifskiWrapper {
 		}
 	}
 
-	func setProgressCallback(
-		context: UnsafeMutableRawPointer,
-		callback: @escaping (@convention(c) (UnsafeMutableRawPointer?) -> Int32)
-	) {
-		gifski_set_progress_callback(pointer, callback, context)
+	typealias ProgressCallback = () -> Int
+
+	private var progressCallback: ProgressCallback!
+
+	func setProgressCallback(_ callback: @escaping ProgressCallback) {
+		progressCallback = callback
+
+		gifski_set_progress_callback(
+			pointer,
+			{ context in // swiftlint:disable:this opening_brace
+				let this = Unmanaged<GifskiWrapper>.fromOpaque(context!).takeUnretainedValue()
+				return Int32(this.progressCallback())
+			},
+			Unmanaged.passRetained(self).toOpaque()
```

Comment:
I have no idea how manual memory management works. It only works when passed as retained here. If I use `. passUnretained()`, it crashes as usual.

@kornelski Do you think this handling is correct? I guess `gifski_set_progress_callback` does not retain the `context` for me?

I have a feeling the problem before was that the passed in `context` got deallocated when Gifski cancelled the conversion, but libgifski still tried to call the callback afterwards, resulting in a crash. Any thoughts on that theory?

Reply:
Yeah, C API doesn't even know retaining exists. Never retains anything.

Reply:
Does passRetained increase refcount? If so, it may be leaking memory, because there won't be anything to decrease the refcount later. But leaks are better than crashes :)

Reply:
Could the C API check whether `context` is currently allocated and only call the callback if it is?

I’ll look into how to decrease the ref count when it’s not longer needed.

Reply:
Unfortunately such thing is not possible. It would be a life-saver for all user-after-free C bugs :) 

The closest approximation would be to take the callback as a weak reference, but C doesn't have them either.

Swift's memory management is unclear to me: it's automatic magic, except where it suddenly isn't :(

Maybe it would help to write an Objective-C wrapper for the C API?